packages/server-core/src/projects/project-history/hooks.ts

```diff
@@ -0,0 +1,245 @@
+        actionType = 'RESOURCE_CREATED'
+      } else if (context.method === 'update' || context.method === 'patch') {
+        const keyBeforeUpdate = (context?.keyBeforeUpdate?.[item.id] || '') as string
+        if (!keyBeforeUpdate) {
+          actionDetail['url'] = item.key
+          actionType = 'RESOURCE_MODIFIED'
+        } else {
+          const newURL = item.key
+          const oldURL = keyBeforeUpdate
+
+          if (newURL !== oldURL) {
+            actionDetail['oldURL'] = oldURL
+            actionDetail['newURL'] = newURL
+            actionType = 'RESOURCE_RENAMED'
+          } else {
+            actionDetail['url'] = item.key
+            actionType = 'RESOURCE_MODIFIED'
+          }
+        }
+      } else {
+        actionDetail['url'] = item.key
+        actionType = 'RESOURCE_REMOVED'
+      }
+
+      if (item.type === 'scene') {
+        actionType = actionType.replace('RESOURCE', 'SCENE') as ActionType
+      }
+
+      const actionDetailStr = JSON.stringify(actionDetail)
+
+      await context.app.service(projectHistoryPath).create({
+        projectId: project.id,
+        userId: context.params.user?.id || null,
+        action: actionType,
+        actionIdentifier: item.id,
+        actionIdentiferType: staticResourcePath,
+        actionDetail: actionDetailStr
+      })
+    }
+  }
+}
+
+const updateProjectHistory = async (context: HookContext<ProjectService>) => {
+  const data = context.result
+  const dataArr = data ? (Array.isArray(data) ? data : 'data' in data ? data.data : [data]) : []
+
+  for (const item of dataArr) {
+    const actionDetail = {
+      projectName: item.name
+    }
+    const actionDetailStr = JSON.stringify(actionDetail)
+
+    await context.app.service(projectHistoryPath).create({
+      projectId: item.id,
+      userId: context.params.user?.id || null,
+      action: 'PROJECT_CREATED',
+      actionIdentifier: item.id,
+      actionIdentiferType: projectPath,
+      actionDetail: actionDetailStr
+    })
+  }
+}
+
+const storeResourceKey = async (context: HookContext<StaticResourceService>) => {
+  context.keyBeforeUpdate = {} as Record<string, string>
+
+  const resources = await context.app.service(staticResourcePath).find({
+    query: {
+      id: context.id as string
+    },
+    paginate: false
+  })
+
+  for (const resource of resources) {
+    context['keyBeforeUpdate'][resource.id] = resource.key
+  }
+}
+
+const storePermissionType = async (context: HookContext<ProjectPermissionService>) => {
+  context.permissionTypeBeforeUpdate = {} as Record<string, string>
+
+  const projectPermissions = await context.app.service(projectPermissionPath).find({
+    query: {
+      id: context.id as string
+    },
+    paginate: false
+  })
+
+  for (const projectPermission of projectPermissions) {
+    context['permissionTypeBeforeUpdate'][projectPermission.id] = projectPermission.type
+  }
+}
+
+staticResourceHooks.before.update.unshift(storeResourceKey)
```

Comment:
we need to move these into the respective hook definitions, rather than injecting them here

Reply:
Since we're discussing whether to follow hooks or trigger approach, I'll pause on this, until we confirm what we should finalise.

Reply:
Removed all hooks